File: kernel/src/repositories/todo.rs
use std::sync::Arc;

use async_trait::async_trait;
use chrono::Utc;
use sea_orm::prelude::Date;
use sea_orm::{
    ActiveModelTrait, ActiveValue::Set, ColumnTrait, DatabaseConnection, EntityTrait,
    IntoActiveModel, QueryFilter,
};
use uuid::Uuid;

use crate::adapters::meta::RequestMeta;
use crate::utils::extract_req_meta;
use crate::{
    adapters::{
        recycle_bin::{CreateRecycleBinEntry, RecycleBinItemType},
        todo::{CreateTodo, TodoPriority, UpdateTodo},
    },
    entities::todo,
    error::KernelError,
    repositories::{
        prelude::WorkspaceRepositoryExt,
        recycle_bin::{RecycleBinRepository, RecycleBinRepositoryExt},
        workspace::WorkspaceRepository,
        workspace_manager::{DuplicateRecord, RecordExistInWorkspace, TransferRecord},
    },
};

#[derive(Debug, Clone)]
pub struct TodoRepository {
    conn: Arc<DatabaseConnection>,
    workspace_repository: WorkspaceRepository,
}

#[async_trait]
pub trait TodoRepositoryExt {
    fn new(conn: Arc<DatabaseConnection>) -> Self;

    async fn create_todo(
        &self,
        payload: &CreateTodo,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError>;

    async fn find_by_id(
        &self,
        identifier: &Uuid,
        meta: &Option<RequestMeta>,
    ) -> Result<Option<todo::Model>, KernelError>;

    async fn find_all(&self, meta: &Option<RequestMeta>) -> Result<Vec<todo::Model>, KernelError>;

    async fn update(
        &self,
        identifier: &Uuid,
        payload: &UpdateTodo,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError>;

    async fn delete(
        &self,
        identifier: &Uuid,
        meta: &Option<RequestMeta>,
    ) -> Result<(), KernelError>;

    async fn change_priority(
        &self,
        identifier: &Uuid,
        priority: &TodoPriority,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError>;

    async fn update_due_date(
        &self,
        identifier: &Uuid,
        due_date: Option<Date>,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError>;

    async fn mark_done(
        &self,
        identifier: &Uuid,
        done: bool,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError>;
}

#[async_trait]
impl TodoRepositoryExt for TodoRepository {
    fn new(conn: Arc<DatabaseConnection>) -> Self {
        Self {
            conn: conn.clone(),
            workspace_repository: WorkspaceRepository::new(conn.clone()),
        }
    }

    async fn create_todo(
        &self,
        payload: &CreateTodo,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError> {
        let mut active_model: todo::ActiveModel = payload.to_owned().into();

        if let Some(meta) = meta {
            active_model.workspace_identifier = Set(Some(meta.workspace_identifier));
        } else {
            return Err(KernelError::DbOperationError(
                "workspace identifier is required".into(),
            ));
        };
        active_model
            .insert(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))
    }

    async fn find_by_id(
        &self,
        identifier: &Uuid,
        meta: &Option<RequestMeta>,
    ) -> Result<Option<todo::Model>, KernelError> {
        let meta = extract_req_meta(meta)?;

        todo::Entity::find()
            .filter(todo::Column::Identifier.eq(*identifier))
            .filter(todo::Column::WorkspaceIdentifier.eq(meta.workspace_identifier))
            .one(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))
    }

    async fn find_all(&self, meta: &Option<RequestMeta>) -> Result<Vec<todo::Model>, KernelError> {
        let meta = extract_req_meta(meta)?;

        todo::Entity::find()
            .filter(todo::Column::WorkspaceIdentifier.eq(meta.workspace_identifier))
            .all(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))
    }

    async fn update(
        &self,
        identifier: &Uuid,
        payload: &UpdateTodo,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError> {
        let meta = extract_req_meta(meta)?;

        let model = todo::Entity::find()
            .filter(todo::Column::WorkspaceIdentifier.eq(meta.workspace_identifier))
            .filter(todo::Column::Identifier.eq(*identifier))
            .one(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?
            .ok_or_else(|| KernelError::DbOperationError("todo not found".to_string()))?;

        let mut active_model = model.into_active_model();

        if let Some(title) = &payload.title {
            active_model.title = Set(title.clone());
        }
        if let Some(description) = &payload.description {
            active_model.description = Set(Some(description.clone()));
        }
        active_model.updated_at = Set(Utc::now().fixed_offset());

        active_model
            .update(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))
    }

    async fn delete(
        &self,
        identifier: &Uuid,
        meta: &Option<RequestMeta>,
    ) -> Result<(), KernelError> {
        let meta = extract_req_meta(meta)?;

        let model = todo::Entity::find()
            .filter(todo::Column::Identifier.eq(*identifier))
            .filter(todo::Column::WorkspaceIdentifier.eq(meta.workspace_identifier))
            .one(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?
            .ok_or_else(|| KernelError::DbOperationError("todo not found".to_string()))?;

        let payload = serde_json::to_string(&model)
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?;

        RecycleBinRepository::new(self.conn.clone())
            .store(
                &CreateRecycleBinEntry {
                    item_id: model.identifier,
                    item_type: RecycleBinItemType::Todo,
                    workspace_identifier: model.workspace_identifier,
                    payload,
                },
                &Some(meta.clone()),
            )
            .await?;

        todo::Entity::delete_many()
            .filter(todo::Column::Identifier.eq(*identifier))
            .exec(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?;
        Ok(())
    }

    async fn change_priority(
        &self,
        identifier: &Uuid,
        priority: &TodoPriority,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError> {
        let meta = extract_req_meta(meta)?;

        let model = todo::Entity::find()
            .filter(todo::Column::Identifier.eq(*identifier))
            .filter(todo::Column::WorkspaceIdentifier.eq(meta.workspace_identifier))
            .one(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?
            .ok_or_else(|| KernelError::DbOperationError("todo not found".to_string()))?;

        let mut active_model = model.into_active_model();
        active_model.priority = Set(priority.to_string());
        active_model.updated_at = Set(Utc::now().fixed_offset());

        active_model
            .update(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))
    }

    async fn update_due_date(
        &self,
        identifier: &Uuid,
        due_date: Option<Date>,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError> {
        let meta = extract_req_meta(meta)?;

        let model = todo::Entity::find()
            .filter(todo::Column::WorkspaceIdentifier.eq(meta.workspace_identifier))
            .filter(todo::Column::Identifier.eq(*identifier))
            .one(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?
            .ok_or_else(|| KernelError::DbOperationError("todo not found".to_string()))?;

        let mut active_model = model.into_active_model();
        active_model.due_date = Set(due_date);
        active_model.updated_at = Set(Utc::now().fixed_offset());

        active_model
            .update(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))
    }

    async fn mark_done(
        &self,
        identifier: &Uuid,
        done: bool,
        meta: &Option<RequestMeta>,
    ) -> Result<todo::Model, KernelError> {
        let meta = extract_req_meta(meta)?;

        let model = todo::Entity::find()
            .filter(todo::Column::WorkspaceIdentifier.eq(meta.workspace_identifier))
            .filter(todo::Column::Identifier.eq(*identifier))
            .one(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?
            .ok_or_else(|| KernelError::DbOperationError("todo not found".to_string()))?;

        let mut active_model = model.into_active_model();
        active_model.done = Set(done);
        active_model.updated_at = Set(Utc::now().fixed_offset());

        active_model
            .update(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))
    }
}

impl TransferRecord for TodoRepository {
    async fn transfer_record(
        &self,
        record_identifier: &Uuid,
        previous_workspace_identifier: &Uuid,
        target_workspace_identifier: &Uuid,
    ) -> Result<(), KernelError> {
        let (prev_exists_res, target_exists_res) = tokio::join!(
            self.workspace_repository
                .exists(previous_workspace_identifier),
            self.workspace_repository
                .exists(target_workspace_identifier),
        );

        let prev_exists = prev_exists_res?;
        let target_exists = target_exists_res?;

        if !prev_exists {
            return Err(KernelError::WorkspaceNotFound(
                previous_workspace_identifier.to_string(),
            ));
        }

        if !target_exists {
            return Err(KernelError::WorkspaceNotFound(
                target_workspace_identifier.to_string(),
            ));
        }

        if !self
            .record_exists_in_workspace(record_identifier, previous_workspace_identifier)
            .await?
        {
            return Err(KernelError::TodoNotFound(record_identifier.to_string()));
        }

        let Some(record) = todo::Entity::find()
            .filter(todo::Column::Identifier.eq(*record_identifier))
            .one(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?
        else {
            return Err(KernelError::TodoNotFound(record_identifier.to_string()));
        };

        let mut active_model = record.into_active_model();

        active_model.updated_at = Set(Utc::now().fixed_offset());
        active_model.workspace_identifier = Set(Some(*target_workspace_identifier));

        active_model
            .update(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?;

        Ok(())
    }
}

impl RecordExistInWorkspace for TodoRepository {
    async fn record_exists_in_workspace(
        &self,
        record_identifier: &Uuid,
        workspace_identifier: &Uuid,
    ) -> Result<bool, KernelError> {
        let record = todo::Entity::find()
            .filter(todo::Column::Identifier.eq(*record_identifier))
            .filter(todo::Column::WorkspaceIdentifier.eq(*workspace_identifier))
            .one(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?;

        Ok(record.is_some())
    }
}

impl DuplicateRecord for TodoRepository {
    async fn duplicate_record(
        &self,
        record_identifier: &Uuid,
        previous_workspace_identifier: &Uuid,
        target_workspace_identifier: &Uuid,
    ) -> Result<(), KernelError> {
        let (prev_exists_res, target_exists_res) = tokio::join!(
            self.workspace_repository
                .exists(previous_workspace_identifier),
            self.workspace_repository
                .exists(target_workspace_identifier),
        );

        let prev_exists = prev_exists_res?;
        let target_exists = target_exists_res?;

        if !prev_exists {
            return Err(KernelError::WorkspaceNotFound(
                previous_workspace_identifier.to_string(),
            ));
        }

        if !target_exists {
            return Err(KernelError::WorkspaceNotFound(
                target_workspace_identifier.to_string(),
            ));
        }

        let Some(record) = todo::Entity::find()
            .filter(todo::Column::Identifier.eq(*record_identifier))
            .filter(todo::Column::WorkspaceIdentifier.eq(*previous_workspace_identifier))
            .one(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?
        else {
            return Err(KernelError::TodoNotFound(record_identifier.to_string()));
        };

        let mut new_record = record.into_active_model();

        new_record.identifier = Set(Uuid::new_v4());
        new_record.workspace_identifier = Set(Some(*target_workspace_identifier));
        new_record.created_at = Set(Utc::now().fixed_offset());
        new_record.updated_at = Set(Utc::now().fixed_offset());

        new_record
            .insert(self.conn.as_ref())
            .await
            .map_err(|err| KernelError::DbOperationError(err.to_string()))?;

        Ok(())
    }
}

File: nuxt-app/appllll/utils/lunar/repositories/todo.ts
import type { CreateTodo, Priority, Todo, UpdateTodo } from "lunar";
import { BaseRepository, type RequestMeta } from "../base";

export type { CreateTodo, UpdateTodo };

const COLUMNS = [
  "title",
  "description",
  "due_date",
  "priority",
  "done",
  "workspace_identifier",
  "created_at",
  "updated_at",
];

export class TodoRepository extends BaseRepository {
  async create_todo(payload: CreateTodo, meta?: RequestMeta): Promise<Todo> {
    const m = this.requireMeta(meta);
    const now = this.now();
    return this.mustRow<Todo>(
      `INSERT INTO todo
         (identifier, title, description, due_date, priority, done, created_at, updated_at, workspace_identifier)
       VALUES ($1, $2, $3, $4, $5, FALSE, $6, $7, $8)
       RETURNING *`,
      [
        this.newUuid(),
        payload.title,
        payload.description ?? null,
        payload.dueDate ?? null,
        payload.priority,
        now,
        now,
        m.workspaceIdentifier,
      ],
    );
  }

  async find_by_id(
    identifier: string,
    meta?: RequestMeta,
  ): Promise<Todo | null> {
    const m = this.requireMeta(meta);
    return this.row<Todo>(
      `SELECT * FROM todo WHERE identifier = $1 AND workspace_identifier = $2`,
      [identifier, m.workspaceIdentifier],
    );
  }

  async find_all(meta?: RequestMeta): Promise<Todo[]> {
    const m = this.requireMeta(meta);
    return this.rows<Todo>(
      `SELECT * FROM todo WHERE workspace_identifier = $1 ORDER BY created_at DESC`,
      [m.workspaceIdentifier],
    );
  }

  async change_priority(
    identifier: string,
    priority: Priority,
    meta?: RequestMeta,
  ): Promise<Todo> {
    const m = this.requireMeta(meta);
    const row = await this.row<Todo>(
      `UPDATE todo SET priority = $2, updated_at = $3
       WHERE identifier = $1 AND workspace_identifier = $4
       RETURNING *`,
      [identifier, priority, this.now(), m.workspaceIdentifier],
    );
    if (!row) throw new Error("todo not found");
    return row;
  }

  async mark_done(
    identifier: string,
    done: boolean,
    meta?: RequestMeta,
  ): Promise<Todo> {
    const m = this.requireMeta(meta);
    const row = await this.row<Todo>(
      `UPDATE todo SET done = $2, updated_at = $3
       WHERE identifier = $1 AND workspace_identifier = $4
       RETURNING *`,
      [identifier, done, this.now(), m.workspaceIdentifier],
    );
    if (!row) throw new Error("todo not found");
    return row;
  }

  async update_due_date(
    identifier: string,
    due_date: string | null,
    meta?: RequestMeta,
  ): Promise<Todo> {
    const m = this.requireMeta(meta);
    const row = await this.row<Todo>(
      `UPDATE todo SET due_date = $2, updated_at = $3
       WHERE identifier = $1 AND workspace_identifier = $4
       RETURNING *`,
      [identifier, due_date, this.now(), m.workspaceIdentifier],
    );
    if (!row) throw new Error("todo not found");
    return row;
  }

  async update(
    identifier: string,
    payload: UpdateTodo,
    meta?: RequestMeta,
  ): Promise<Todo> {
    const m = this.requireMeta(meta);
    const sets: string[] = ["updated_at = $2"];
    const params: unknown[] = [identifier, this.now()];
    let idx = 3;

    if (payload.title !== undefined) {
      sets.push(`title = $${idx++}`);
      params.push(payload.title);
    }
    if (payload.description !== undefined) {
      sets.push(`description = $${idx++}`);
      params.push(payload.description);
    }

    const row = await this.row<Todo>(
      `UPDATE todo SET ${sets.join(", ")} WHERE identifier = $1 AND workspace_identifier = $${idx}
       RETURNING *`,
      [...params, m.workspaceIdentifier],
    );
    if (!row) throw new Error("todo not found");
    return row;
  }

  async delete(identifier: string, meta?: RequestMeta): Promise<void> {
    const m = this.requireMeta(meta);
    const model = await this.find_by_id(identifier, m);
    if (!model) throw new Error("todo not found");

    await this.run(
      `INSERT INTO recycle_bin
         (identifier, item_id, item_type, payload, deleted_at, workspace_identifier)
       VALUES ($1, $2, 'todo', $3, $4, $5)`,
      [
        this.newUuid(),
        identifier,
        JSON.stringify(model),
        this.now(),
        m.workspaceIdentifier,
      ],
    );

    await this.run(
      `DELETE FROM todo WHERE identifier = $1 AND workspace_identifier = $2`,
      [identifier, m.workspaceIdentifier],
    );
  }

  async transfer_record(
    record_identifier: string,
    previous_workspace_identifier: string,
    target_workspace_identifier: string,
  ): Promise<void> {
    await this.transferWorkspaceRecord(
      "todo",
      record_identifier,
      previous_workspace_identifier,
      target_workspace_identifier,
    );
  }

  async duplicate_record(
    record_identifier: string,
    previous_workspace_identifier: string,
    target_workspace_identifier: string,
  ): Promise<void> {
    await this.duplicateWorkspaceRecord(
      "todo",
      record_identifier,
      previous_workspace_identifier,
      target_workspace_identifier,
      COLUMNS,
    );
  }

  async record_exists_in_workspace(
    record_identifier: string,
    workspace_identifier: string,
  ): Promise<boolean> {
    return this.recordExistsInWorkspace(
      "todo",
      record_identifier,
      workspace_identifier,
    );
  }
}
